cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -237,6 +237,23 @@ class ParquetScanTaskIterator {
   std::shared_ptr<parquet::arrow::FileReader> reader_;
 };
 
+Status ParquetFileFormat::IsSupported(const FileSource& source, bool* supported) const {
+  try {
+    std::shared_ptr<io::RandomAccessFile> input;
+    RETURN_NOT_OK(source.Open(&input));
+    auto reader = parquet::ParquetFileReader::Open(input);
+  } catch (::parquet::ParquetInvalidOrCorruptedFileException) {
```

Comment:
```suggestion
  } catch (const ::parquet::ParquetInvalidOrCorruptedFileException&) {
```

Reply:
(this is what's causing your StringBuilder instantiation failure- this catch declaration is trying to construct a new instance of `::parquet::ParquetInvalidOrCorruptedFileException` using the variadic tempate constructor:
```c++
  } catch (::parquet::ParquetInvalidOrCorruptedFileException e(__caught)) {
```
)

You might also consider explicitly defaulting/deleting default/copy/move constructors for `::parquet::ParquetInvalidOrCorruptedFileException`.

Reply:
I changed it due to msvc complaining about `e` not being used.

Reply:
I'll just use a plain ol' `(void)e;`.

Reply:
The current error is about catching the exception by value rather than by const reference

Reply:
It's because I changed it.